merge 'simple-jobs' and 'static-dashboard' into 'master'

jsnal opened the pull request:
Merge all changes from 'simple-job' and 'static-dashboard'

Change log:

- Static front page with HTML cards updating on an interval
- The cards are dynamically created from the nodes in the cluster
- Statistics are formatted as percents out of 100
- Simple job submission API
- Send formatted text from a HTML textarea to the Flask webserver
- Save text to a file within a directory picked by the user
- Run file as a SLURM batch job

jsnal commented on src/env.py:
```diff
@@ -0,0 +1,21 @@
+CLUSTER_LOG_PATH  = os.getenv('CLUSTER_LOG_PATH')  if os.getenv('CLUSTER_LOG_PATH')  else f"{CLUSTER_SHARED_FS}/cluster_logs"
+
+
+def _get_init_job_id():
```
This gets the ID by looking at the sub directories and grabbing the highest number. For example if the directories are 1, 2, 3, 4; the current ID would be 4.

jsnal commented on src/slurm/sbatch.py:
```diff
@@ -0,0 +1,11 @@
+
+def sbatch_run(filepath, *args, **kwargs):
+    batch_command = f'sbatch --output={CLUSTER_LOG_PATH}/{get_job_id()}/{get_job_id()}.out {filepath}'
+    os.system(batch_command)
```
This is still a work in progress. Soon we will be adding the ability to change all of the different options within the jobs. Things like node count, core count, timeouts, etc.

paulcsmith replied:
👍

jsnal commented on src/utils/crlf_to_lf.py:
```diff
@@ -0,0 +1,10 @@
+        content = content.replace(WINDOWS_LINE_ENDING, UNIX_LINE_ENDING)
+
+    with open(file_path, 'wb') as open_file:
+        open_file.write(content)
```
Replaces windows '\r\n' generated by the HTML with UNIX '\n'.

paulcsmith replied:
I'd consider adding a comment in the code or extracting a function called `convert_windows_line_endings_to_unix` so the function itself is self documenting

paulcsmith replied:
Actually could you use https://github.com/CamJW101/rpi-cluster-controller/pull/2/files#diff-e59babf30702c5551b2534e7a489e22fR18? If not that's fine, just making sure

paulcsmith commented on src/env.py:
```diff
@@ -0,0 +1,21 @@
+def _get_init_job_id():
+    try:
+        return int(max(next(os.walk(CLUSTER_LOG_PATH))[1]))
+    except:
```
I would recommend catching a specific error https://stackoverflow.com/questions/13531247/python-catching-specific-exception

The reason for that is that if at some point you have a legit bug on line 9 and you don't catch a specific error then you may not even realize there's a bug or where the problem is coming for. Usually the more specific the better when catching errors.

paulcsmith commented on src/job.py:
```diff
@@ -0,0 +1,42 @@
+        # Run the new job as a batch job
+        sbatch_run(full_file_path)
+
+    except Exception:
```
Same here with catching a specific error (or error message). You may also want touse the error message  in the json dump to make it easier to debug from the client side

paulcsmith replied:
Though if you really do want to just catch everything a nd return an error JSON object then you can do that, but I'd recommend printing the  exception message and maybe including it in the JSON so it doesn't get totally swallowed

paulcsmith commented on src/utils/async_request.py:
```diff
@@ -4,9 +4,13 @@
+        print('Request Timed Out:', url)
+        return None
+    except Exception as e:
+        print('Exception in Request:', url, e)
```
Nice! That should help with debugging 👍

jsnal commented:
Changed some of the things you mentioned in your review. I wasn't exactly sure what you meant by the error catching in src/job.py. I just returned the error message in the json string and printed it to the console so you can see where the code failed.

I also renamed crlf_to_lf to convert_windows_line_endings_to_unix like you recommended.

paulcsmith commented on src/job.py:
```diff
@@ -0,0 +1,48 @@
+        sbatch_run(full_file_path)
+    except Exception as e:
+        print('Failed running the batch job:', e)
+        return json.dumps({'status': 'failed running job', 'code': e})
```
Looks good!
